probe-rs-tools/src/bin/probe-rs/util/rtt/processing.rs

```diff
@@ -62,14 +62,13 @@ impl RttDecoder {
         matches!(self, RttDecoder::BinaryLE)
     }
 
-    pub fn process(
-        &mut self,
-        buffer: &[u8],
-        collector: &mut impl RttDataHandler,
-    ) -> Result<(), Error> {
+    pub fn process<H>(&mut self, buffer: &[u8], collector: &mut H) -> Result<H::Output, Error>
```

Comment:
Do you have a plan for this Output? H is supposed to be a callback-like type, I'm not sure if a return value makes sense in this abstraction.

Reply:
This is how I get output back in the `impl RttDataHandler for util::cli::Printer` case -- getting it out rather than comprehensively handling it in a callback allows me to work without converting *all* the processing callbacks into async functions. (And even then I'd have to turn the single Printer into a printer factory that takes the shortlived &mut to the file into the callbacks on demand).

It's clearly doable without taking the return-data route, and yes probably cleaner, but more (hopefully internal) API churn. (Then again, I already had to switch all the `on_msg` to async functions, which conveniently have become usable just a release ago).

Reply:
You could also push the output into a channel, and process them on the outside. I don't have strong objections here, but I'm also not a big fan of the change.

Reply:
Sure I am!

Reply:
As this is now rebased onto #3254, this hack could go away: instead, a short-lived printer is constructed for every processing event (out of the prefix for showing, and an `Option<&mut>` to the file), and prints both to the display and to the file.